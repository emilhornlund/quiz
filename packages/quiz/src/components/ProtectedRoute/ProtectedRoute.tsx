import React, { FC } from 'react'
import { Navigate } from 'react-router-dom'

import { useAuthContext } from '../../context/auth'

export interface ProtectedRouteProps {
  children: React.ReactNode
}

const ProtectedRoute: FC<ProtectedRouteProps> = ({ children }) => {
  const auth = useAuthContext()

  if (!auth.isLoggedIn) {
    return <Navigate to="/" replace />
  }

  return children
}

export default ProtectedRoute
