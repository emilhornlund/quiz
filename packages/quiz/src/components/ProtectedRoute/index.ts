export type { ProtectedRouteProps } from './ProtectedRoute'
export { default } from './ProtectedRoute'
