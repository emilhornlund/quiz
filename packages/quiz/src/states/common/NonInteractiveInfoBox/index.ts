export type { NonInteractiveInfoBoxProps } from './NonInteractiveInfoBox'
export { default } from './NonInteractiveInfoBox'
