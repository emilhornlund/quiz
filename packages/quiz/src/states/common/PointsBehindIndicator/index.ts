export type { PointBehindIndicatorProps } from './PointsBehindIndicator'
export { default } from './PointsBehindIndicator'
