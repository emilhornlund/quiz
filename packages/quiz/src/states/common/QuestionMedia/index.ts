export type { QuestionMediaProps } from './QuestionMedia'
export { default } from './QuestionMedia'
