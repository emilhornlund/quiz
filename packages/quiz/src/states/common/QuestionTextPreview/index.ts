export type { QuestionTextPreviewProps } from './QuestionTextPreview'
export { default } from './QuestionTextPreview'
