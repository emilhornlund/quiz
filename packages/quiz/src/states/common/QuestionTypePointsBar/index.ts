export type { QuestionTypePointsBarProps } from './QuestionTypePointsBar'
export { default } from './QuestionTypePointsBar'
