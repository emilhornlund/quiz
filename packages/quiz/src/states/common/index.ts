export type { NonInteractiveInfoBoxProps } from './NonInteractiveInfoBox'
export type { PointBehindIndicatorProps } from './PointsBehindIndicator'
export type { QuestionAnswerPickerProps } from './QuestionAnswerPicker'
export type { QuestionMediaProps } from './QuestionMedia'
export type { QuestionTextPreviewProps } from './QuestionTextPreview'
export type { QuestionTypePointsBarProps } from './QuestionTypePointsBar'

export { default as GamePage } from './GamePage'
export { default as NonInteractiveInfoBox } from './NonInteractiveInfoBox'
export { default as PointsBehindIndicator } from './PointsBehindIndicator'
export { default as QuestionAnswerPicker } from './QuestionAnswerPicker'
export { default as QuestionMedia } from './QuestionMedia'
export { default as QuestionTextPreview } from './QuestionTextPreview'
export { default as QuestionTypePointsBar } from './QuestionTypePointsBar'
