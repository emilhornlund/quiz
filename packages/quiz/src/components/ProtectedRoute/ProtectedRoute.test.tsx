import { TokenScope } from '@quiz/common'
import { render, screen } from '@testing-library/react'
import React from 'react'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import ProtectedRoute from './ProtectedRoute'

// --- Mock AuthContext hook with mutable state ---
type MockAuth = { isUserAuthenticated: boolean; isGameAuthenticated: boolean }
let mockAuth: MockAuth = {
  isUserAuthenticated: false,
  isGameAuthenticated: false,
}

vi.mock('../../context/auth', () => ({
  useAuthContext: () => mockAuth,
}))

function renderAt(path: string, ui: React.ReactNode) {
  return render(
    <MemoryRouter initialEntries={[path]}>
      <Routes>
        <Route path="/" element={<div>Home</div>} />
        <Route path="/protected" element={ui} />
      </Routes>
    </MemoryRouter>,
  )
}

describe('ProtectedRoute', () => {
  beforeEach(() => {
    mockAuth = { isUserAuthenticated: false, isGameAuthenticated: false }
  })

  it('renders children when User scope is authenticated (default props)', () => {
    mockAuth.isUserAuthenticated = true

    const { container } = renderAt(
      '/protected',
      <ProtectedRoute>
        <div>Protected User</div>
      </ProtectedRoute>,
    )

    expect(screen.getByText('Protected User')).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('redirects to "/" when User scope is not authenticated (default authenticated=true)', () => {
    mockAuth.isUserAuthenticated = false

    const { container } = renderAt(
      '/protected',
      <ProtectedRoute>
        <div>Protected User</div>
      </ProtectedRoute>,
    )

    expect(screen.getByText('Home')).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('renders children for guest-only route when User is NOT authenticated (authenticated=false)', () => {
    mockAuth.isUserAuthenticated = false

    const { container } = renderAt(
      '/protected',
      <ProtectedRoute authenticated={false}>
        <div>Guest Only</div>
      </ProtectedRoute>,
    )

    expect(screen.getByText('Guest Only')).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('redirects for guest-only route when User IS authenticated (authenticated=false)', () => {
    mockAuth.isUserAuthenticated = true

    const { container } = renderAt(
      '/protected',
      <ProtectedRoute authenticated={false}>
        <div>Guest Only</div>
      </ProtectedRoute>,
    )

    expect(screen.getByText('Home')).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('renders children when Game scope is authenticated', () => {
    mockAuth.isGameAuthenticated = true

    const { container } = renderAt(
      '/protected',
      <ProtectedRoute scope={TokenScope.Game}>
        <div>Protected Game</div>
      </ProtectedRoute>,
    )

    expect(screen.getByText('Protected Game')).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('redirects when Game scope is NOT authenticated', () => {
    mockAuth.isGameAuthenticated = false

    const { container } = renderAt(
      '/protected',
      <ProtectedRoute scope={TokenScope.Game}>
        <div>Protected Game</div>
      </ProtectedRoute>,
    )

    expect(screen.getByText('Home')).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })
})
