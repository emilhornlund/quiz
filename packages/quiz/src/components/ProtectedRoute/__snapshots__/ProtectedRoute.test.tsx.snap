// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProtectedRoute > redirects for guest-only route when User IS authenticated (authenticated=false) 1`] = `
<div>
  <div>
    Home
  </div>
</div>
`;

exports[`ProtectedRoute > redirects to "/" when User scope is not authenticated (default authenticated=true) 1`] = `
<div>
  <div>
    Home
  </div>
</div>
`;

exports[`ProtectedRoute > redirects when Game scope is NOT authenticated 1`] = `
<div>
  <div>
    Home
  </div>
</div>
`;

exports[`ProtectedRoute > renders children for guest-only route when User is NOT authenticated (authenticated=false) 1`] = `
<div>
  <div>
    Guest Only
  </div>
</div>
`;

exports[`ProtectedRoute > renders children when Game scope is authenticated 1`] = `
<div>
  <div>
    Protected Game
  </div>
</div>
`;

exports[`ProtectedRoute > renders children when User scope is authenticated (default props) 1`] = `
<div>
  <div>
    Protected User
  </div>
</div>
`;
